infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java

```diff
@@ -100,21 +132,72 @@ protected KubernetesEnvironment(
     super(internalRecipe, machines, warnings);
     setType(TYPE);
     this.pods = pods;
+    this.deployments = deployments;
     this.services = services;
     this.ingresses = ingresses;
     this.persistentVolumeClaims = persistentVolumeClaims;
     this.secrets = secrets;
     this.configMaps = configMaps;
+    this.podData = new HashMap<>();
+    pods.entrySet()
+        .forEach(
+            e -> {
+              Pod pod = e.getValue();
+              podData.put(e.getKey(), new PodData(pod.getSpec(), pod.getMetadata()));
+            });
+    deployments
+        .entrySet()
+        .forEach(
+            e -> {
+              PodTemplateSpec podTemplate = e.getValue().getSpec().getTemplate();
+              podData.put(
+                  e.getKey(), new PodData(podTemplate.getSpec(), podTemplate.getMetadata()));
+            });
   }
 
   @Override
   public KubernetesEnvironment setType(String type) {
     return (KubernetesEnvironment) super.setType(type);
   }
 
-  /** Returns pods that should be created when environment starts. */
-  public Map<String, Pod> getPods() {
-    return pods;
+  /**
+   * Returns pods that should be created when environment starts.
+   *
+   * <p>Note: This map <b>should not</b> be changed, as it will only return pods and not
+   * deployments. If objects in the map need to be changed, see {@link #getPodsData()}
+   */
+  public Map<String, Pod> getPodsCopy() {
+    return ImmutableMap.copyOf(pods);
+  }
+
+  /**
+   * Returns deployments that should be created when environment starts.
+   *
+   * <p>Note: This map <b>should not</b> be changed. If objects in the map need to be changed, see
+   * {@link #getPodsData()}
+   */
+  public Map<String, Deployment> getDeploymentsCopy() {
+    return ImmutableMap.copyOf(deployments);
+  }
+
+  /**
+   * Returns {@link PodData} representing the metadata and pod spec of objects (pods or deployments)
+   * that should be created when environment starts. The data returned by this method represents all
+   * deployment and pod objects that form the workspace, and should be used when provisioning or
+   * performing any action that needs to see every object in the environment.
+   */
+  public Map<String, PodData> getPodsData() {
+    return ImmutableMap.copyOf(podData);
+  }
+
+  /**
+   * Add a pod to the current environment. This method is necessary as the map returned by {@link
+   * #getPodsCopy()} is a copy. This method also adds the relevant data to {@link #getPodsData()}.
+   */
+  public void addPod(Pod pod) {
```

Comment:
It shouldn't be a huge issue, and might make the deploy code simpler, but would require rewriting all of the recipes that use pods to use deployments, and modifying the semantics of converting e.g. docker recipes to deployment objects instead of pods. 

Definitely something we should look at though, since this could clean up a fair bit of the deleting workspaces code I think.